From c314adf8f5755e2fe4cf7fb5720e634e2d5bf48a Mon Sep 17 00:00:00 2001
From: Eskil Abrahamsen Blomfeldt <eskil.abrahamsen-blomfeldt@qt.io>
Date: Fri, 5 Sep 2025 10:34:08 +0200
Subject: [PATCH 007/553] Fix possible assert when inserting columns in
 QTextDocument

In certain tables which have cells that contain row spans
in beginning of a row, we could get into an inconsistent state
which would leave cells in the table unmapped/invalid.

Specifically, this happened when inserting a new left-most
column.

The algorithm has two data structures: A "logical cells map"
which gives the row span and col span of each cell, and then
a grid which maps the logical index of each (x,y) to a cell.

What happened in this case was that we were inserting at
pos == 0. For each row, we would then search for an insertion
point in that row by looking at the grid and finding the
logical index of the first slot in this particular row of the
grid. Then we would insert a new cell directly before this
index in cells and continue.

If we find a logical index which actually belongs to a prior
row than the one we are inserting into, since a previous cell
has a row span, we need to continue searching. Otherwise we
end up inserting the new column into this previous row,
which we have already handled.

This is where it went wrong: There was logic to handle this
by checking the logical index we found against the logical
index of the last grid slot in the previous row, and it is
even documented in a comment. But when pos==0, then we
would *always* set "logicalGridPositionBeforePosition" to -1,
thus we would never actually apply this condition for pos==0.

Setting logicalGridPositionBeforePosition to -1 is correct for
the first row, since there is no previous grid slot to look
at, but for subsequent rows we can still do this look up even
for pos == 0.

Pick-to: 6.9 6.8 6.5
Fixes: QTBUG-138678
Change-Id: I6094e7d0f41611b6d8b0952eac876bf2e41dfe9c
Reviewed-by: Eirik Aavitsland <eirik.aavitsland@qt.io>
(cherry picked from commit 68a5ec23f683c7778f587842fe35a47b545cf9d1)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/gui/text/qtexttable.cpp                   |  2 +-
 .../gui/text/qtexttable/tst_qtexttable.cpp    | 45 +++++++++++++++++++
 2 files changed, 46 insertions(+), 1 deletion(-)

diff --git a/src/gui/text/qtexttable.cpp b/src/gui/text/qtexttable.cpp
index 683ad8ac352..ff8644b5302 100644
--- a/src/gui/text/qtexttable.cpp
+++ b/src/gui/text/qtexttable.cpp
@@ -724,7 +724,7 @@ void QTextTable::insertColumns(int pos, int num)
         if (i == d->nRows - 1 && pos == d->nCols) {
             cell = d->fragment_end;
         } else {
-            int logicalGridIndexBeforePosition = pos > 0
+            int logicalGridIndexBeforePosition = pos > 0 || i > 0
                                                  ? d->findCellIndex(d->grid[i*d->nCols + pos - 1])
                                                  : -1;
 
diff --git a/tests/auto/gui/text/qtexttable/tst_qtexttable.cpp b/tests/auto/gui/text/qtexttable/tst_qtexttable.cpp
index a7e319ef62e..f7e0f788f17 100644
--- a/tests/auto/gui/text/qtexttable/tst_qtexttable.cpp
+++ b/tests/auto/gui/text/qtexttable/tst_qtexttable.cpp
@@ -85,6 +85,8 @@ private slots:
     void columnWidthWithImage();
 #endif
 
+    void QTBUG138678_insertColumnAtStartWithRowspan();
+
 private:
     QTextTable *create2x2Table();
     QTextTable *create4x4Table();
@@ -1413,6 +1415,49 @@ void tst_QTextTable::columnWidthWithImage()
 }
 #endif
 
+void tst_QTextTable::QTBUG138678_insertColumnAtStartWithRowspan()
+{
+    QString s =
+"<html><body>"
+"<table border=1>"
+"<tr>"
+"    <td />"
+"    <td />"
+"    <td />"
+"    <td />"
+"    <td />"
+"</tr>"
+"<tr>"
+"    <td rowspan=2 />"
+"    <td rowspan=2 />"
+"    <td />"
+"    <td />"
+"    <td rowspan=2 />"
+"</tr>"
+"<tr>"
+"    <td />"
+"    <td />"
+"</tr>"
+"</table></body></html>)";
+
+    QTextDocument doc;
+    doc.setHtml(s);
+
+    QTextEdit textEdit;
+    textEdit.setDocument(&doc);
+    textEdit.show();
+    QVERIFY(QTest::qWaitForWindowExposed(&textEdit));
+
+    QTextCursor cursor(doc.firstBlock());
+    cursor.movePosition(QTextCursor::Right);
+
+    QTextTable *currentTable = cursor.currentTable();
+    QVERIFY(currentTable);
+
+    // Don't assert
+    currentTable->insertColumns(0, 1);
+}
+
 
 QTEST_MAIN(tst_QTextTable)
 #include "tst_qtexttable.moc"
-- 
2.51.2

